### ISSUE TITLE
Talkback double announcement when using accessibilityRole={'header'}

### ISSUE THREAD

**Issue by JCron245** [facebook/react-native#28800]
Please provide all the information requested. Issues that do not follow this format are likely to stall.

## Description
For whatever reason when attaching `accessibilityRole={'header'}` to an element it is announcing 'heading' twice with TalkBack on. Testing the same component on iOS using VoiceOver achieves the expected result of it only announcing once.

## React Native version:
61.4

## Steps To Reproduce
Provide a detailed list of steps that reproduce the issue.

1. Create some sort of component with 'accessibilityRole={'header'}` 
2. Test on iOS and Android

## Expected Results
Talkback should announce 'header' only once when using TalkBack like it does with VoiceOver.

## Snack, code example, screenshot, or link to a repository:
https://snack.expo.io/@jcron/accessibility-role-double-announce

**Comment by JCron245** [comment]
Just as additional info I tested the above snack with the following devices:

iPhone X [iOS v13.4] - Works as expected
iPhone XS [iOS v13.5-b3] - Works as expected
iPad Mini 2 [iOS v12.4.5] - Works as expected
NVIDIA Shield Tablet [Android 7.0] - Announces 'heading' Twice
Galaxy Tab 10 [Android 8.1.0] - Announces 'heading' Twice
Moto E5 [Android 8.1.0] - Announces 'heading' Twice

**Comment by AnnieFisher** [comment]
*** Note : 
When you do NOT apply an accessible label it announces "heading {Text} heading".
Not sure why this is or if it is even useful, just additional information.

**Comment by dotjay** [comment]
Another test result: 
**Motorola One Macro XT2016-1** [Android 9.0] - TalkBack announces 'heading' Twice

**Comment by babylone-star** [comment]
This happens on my project, too.

It does not happen when using the "RN Tester" app, which is interesting.

**Comment by babylone-star** [comment]
This has been fixed in v0.63.1:
https://github.com/facebook/react-native/commit/7428271995adf21b2b31b188ed83b785ce1e9189

**Comment by elfstyle** [comment]
Just updated from RN 62.2 to 63.4 and accessibilityRole='header' doesn't work anymore. Does anyone experience the same issue?

**Comment by nico22utn** [comment]
> Just updated from RN 62.2 to 63.4 and accessibilityRole='header' doesn't work anymore. Does anyone experience the same issue?

Yes! I have the same issue for Samsung s10e. It works as expected on any IOS device.

**Comment by amarlette** [comment]
@blavalla has not been able to reproduce this issue on the current RN version. Was this issue resolved by updates since the issue was created?

**Comment by blavalla** [comment]
@arsimr16, this one may actually be caused by a difference in Talkback or Android versions between devices as well, or could possibly be caused by Samsung's own fork of Talkback that they use.

For what it's worth, I wasn't able to reproduce this on a Samsung Galaxy S20 or a Pixel 5. Both were running Android 11 however, and the latest versions of Talkback (Samsung's is technically a fork, but it's a new fork that should be up to date with Google's implementation).

This was the link used to test: https://snack.expo.io/7tjnjlJbO

On a technical note, the "header" role in RN right now is mapping to the older-style of heading that was only allowed within collection structures in Android (lists, grids, etc.). This works on both older and newer OS and Talkback versions (assuming your element is in a list), but starting in Android 9, Google added a more general-purpose API for headings.

Since you are already working with a fork of RN, one thing to try would be to update this to the newer-style of API to see if it works to resolve the issue.  It will limit support down to only API 28 and above (Android 9) or API 19 and above (Android 4.4.) if using the AndroidX compat library.

To do that, you'll want to change these lines:
https://github.com/facebook/react-native/blob/dfa8eb0558338f18ea01f294a64d355f6deeff06/ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java#L398-L400

To this:

`nodeInfo.setHeading(true)`

@kacieb, what is the lowest version of Android that React Native currently supports?  If it's above 19, we may want to just switch to use the newer API here.

**Comment by arsimr16** [comment]
Thanks for the explanation.  I switched to the newer API (I'm using AndroidX) and I still don't hear the header role being announced on my Samsung Galaxy S8 running Android 9.

**Comment by blavalla** [comment]
@arsimr16, just to confirm here. Are you using Google's Talkback (installed from the Play Store), or Samsung's Voice Assistant (comes pre-installed).  Voice Assistant was quite an old fork of talkback, and may not have support for the header role.

If you're using Talkback, which version?

**Comment by arsimr16** [comment]
Thanks, I didn't know that.  I was using the Samsung Voice Assistant.  Let me try downloading Talkback.

**Comment by arsimr16** [comment]
Works with Talkback!  Thanks so much :)

**Comment by kacieb** [comment]
> @kacieb, what is the lowest version of Android that React Native currently supports? If it's above 19, we may want to just switch to use the newer API here.

That's a good idea. The minimum SDK version supported is 21 so we can switch to the newer API.

**Comment by anup690** [comment]
I had an issue with accessibilityRole="link", it's also announced twice like: "<content> one link, link"
anyone had this issue and is it solved after the update?

**Comment by blavalla** [comment]
@anup690 , was this double-link announcement on iOS or Android? If Android, was it using Samsung's "Voice Assistant" or Google's "Talkback"?

If I had to guess, I'd say that this sounds like something that Samsung's Voice Assistant would do. Generally speaking, Voice Assistant tried to be a bit more like Apple's VoiceOver (which automatically announces "link" after links), and RN adding a roleDescription of "Link" may have doubled this up inadvertently. That's a random guess though, and since Samsung has deprecated Voice Assistant in favor of Talkback, I can't really verify it on any of my devices.

If you do have a repro case, see if commenting out this line fixes it:
https://github.com/facebook/react-native/blob/da899c0cc4372830e5ca053a096b74fff2a19cb8/ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java#L375

**Comment by anup690** [comment]
@blavalla you guessed it bang on, it is Samsung's _Voice Assistant_ indeed, got my tester do away with android talkback, yes I recently learned that _Voice Assistant_ is a fork of talkback and outdated,
and the links do read fine on the stock based ROMs (due to Talkback that is)

though with talkback, it reads, `"link", "links available, use swipe up and right to view"` for each link in the page not the best UX,
i'm checking if there is a way to polish that like `"links available, use swipe up and right to view"` to utter only once in a page full of links,
thanks for the help

**Comment by blavalla** [comment]
> i'm checking if there is a way to polish that like `"links available, use swipe up and right to view"` to utter only once in a page full of links,

From an Android perspective, Talkback is doing this for each TextView that contains a ClickableSpan or URLSpan in its text.

From the RN side, whenever you set a link role on a component, we are under the hood creating a URLSpan and setting it as the full length of text on that component:

https://github.com/facebook/react-native/blob/da899c0cc4372830e5ca053a096b74fff2a19cb8/ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java#L377-L387

This is definitely not a great experience currently, and I'm actually working right now to implement a better approach to link handling in general, where the links act more like they do on web surfaces and get focused on their own rather than being accessed via a dedicated menu.
